qubes.lua.in

```diff
@@ -170,60 +161,61 @@ function qubes.get_colour_focus(c)
     return get_colour_by_state(c, 'qubes_label_color_focus')
 end
 
-function qubes.manager_menuitem()
-    return {'Qubes Manager', 'qubes-manager',
-        '/usr/share/icons/hicolor/16x16/apps/qubes-logo-icon.png'}
+function is_accessible(path)
+    return os.rename(path, path) and true or false
 end
 
 function qubes.make_menuitem(vm, program)
-    return {program.Name, program.cmdline, program.icon_path}
+    local appicon = program.icon_path
+    if not is_accessible(appicon) then appicon = nil end
+
+    return {program.Name, program.cmdline, appicon}
 end
 
-function qubes.make_vm_menu(vm, menuitemfunc)
+--all parameters except for vmname are optional
+function qubes.make_vm_menu(vmname, menuitemfunc, vmapps, vmicon)
     local menu = {}
+    local vmapps = vmapps or HOME .. '/.local/share/qubes-appmenus/' .. vmname .. '/apps/'
+    local menuitemfunc = menuitemfunc or qubes.make_menuitem
+    local vmicon = vmicon or '/var/lib/qubes/appvms/' .. vmname .. '/icon.png'
 
-    for _, program in ipairs(menubar.utils.parse_dir(vm.path .. '/apps')) do
-        local menuitem = menuitemfunc(vm, program)
-        if menuitem ~= nil then
-            table.insert(menu, menuitem)
-        end
+    if not is_accessible(vmapps) then return nil end
+
+    for _, program in ipairs(menubar.utils.parse_dir(vmapps)) do
+        local menuitem = menuitemfunc(vmname, program)
+        if menuitem ~= nil then table.insert(menu, menuitem) end
     end
-    return {vm.name, menu, vm.path .. '/icon.png'}
+
+    --NOTE: awesome will not display menu entries with invalid icon paths
+    return {vmname, menu, vmicon}
 end
 
-function qubes.make_menu_with_filter(filter, menuitemfunc)
+--filter & menuitemfunc are optional parameters
+function qubes.make_menu(filter, menuitemfunc)
     local menu = {}
-    for line in io.popen([[python -c "
-import os.path
-import qubes.qubes
-qvmc = qubes.qubes.QubesVmCollection()
-qvmc.lock_db_for_reading()
-qvmc.load()
-qvmc.unlock_db()
-print('\n'.join('{} {} {}'.format(vm.name, vm.type, vm.dir_path)
-    for vm in sorted(qvmc.values(), key=lambda vm: vm.name)
-    if os.path.isdir(vm.dir_path)))
+
+    --set reasonable defaults
+    local menufunc = qubes.make_vm_menu
+    if filter ~= nil then menufunc = filter(vmname) end
+    local menuitemfunc = menuitemfunc or qubes.make_menuitem
+
+    --add the dom0 menu
+    --this is enitrely optional and enables users to create xdg desktop files at the below file path in order to populate the dom0 menu
+    --it will remain invisible until that directory is accessible
+    table.insert(menu, qubes.make_vm_menu('Dom0', nil, HOME .. '/.config/awesome/dom0-menu/', '/usr/share/icons/hicolor/16x16/apps/qubes-logo-icon.png'))
+
+    --add the VM application menus
+    for vmname in io.popen([[python -c "
+from qubesadmin import Qubes
+app = Qubes()
+ret = []
+for vm in app.domains: ret.append(vm.name)
+ret.sort(key=lambda i: i.lower())
+print('\n'.join(ret))
```

Comment:
Better use `qvm-ls --raw-data -O name` to not duplicate that code.
Or `qvm-ls --raw-list`, as soon as I'll bring it back to Qubes 4.0. This way it will work for both 3.2 and 4.0.

Reply:
I could also use a straightforward ls `~/.local/share/qubes-appmenus/`. Or include some library to be able to read directory contents...
But I'll wait for your call wrt the overall future approach.